// @shopify/oxygen-cli/dist/deploy/build-cancel.js
import { graphqlRequest } from '@shopify/cli-kit/node/api/graphql';
import { AbortError } from '@shopify/cli-kit/node/error';
import { outputInfo } from '@shopify/cli-kit/node/output';
import { Header, errorHandler } from '../utils/utils.js';
import { BuildCancelQuery } from './graphql/build-cancel.js';

async function buildCancel(options) {
  const { config, buildId, reason, logger } = options;
  const variables = {
    buildId,
    reason
  };
  try {
    const response = await graphqlRequest({
      query: BuildCancelQuery,
      api: "Oxygen",
      url: `${config.deploymentUrl}/api/v2/admin/graphql`,
      token: config.deploymentToken.accessToken,
      addedHeaders: {
        [Header.OxygenNamespaceHandle]: config.deploymentToken.namespace
      },
      variables
    });
    if (response.buildCancel.userErrors.length >= 1) {
      throw new AbortError(
        `Failed to cancel build: ${response.buildCancel.userErrors[0]?.message}`
      );
    }
    outputInfo(`Build with id ${buildId} cancelled.`, logger);
    return response.buildCancel;
  } catch (error) {
    errorHandler(error);
    throw error;
  }
}

export { buildCancel };
