// @shopify/oxygen-cli/dist/deploy/deployment-cancel.js
import { graphqlRequest } from '@shopify/cli-kit/node/api/graphql';
import { AbortError } from '@shopify/cli-kit/node/error';
import { outputInfo } from '@shopify/cli-kit/node/output';
import { Header, errorHandler } from '../utils/utils.js';
import { DeploymentCancelQuery } from './graphql/deployment-cancel.js';

async function deploymentCancel(options) {
  const { config, deploymentId, reason, logger } = options;
  const variables = {
    deploymentId,
    reason
  };
  try {
    const response = await graphqlRequest({
      query: DeploymentCancelQuery,
      api: "Oxygen",
      url: `${config.deploymentUrl}/api/v2/admin/graphql`,
      token: config.deploymentToken.accessToken,
      variables,
      addedHeaders: {
        [Header.OxygenNamespaceHandle]: config.deploymentToken.namespace
      }
    });
    if (response.deploymentCancel.userErrors.length >= 1) {
      throw new AbortError(
        response.deploymentCancel.userErrors[0]?.message || ""
      );
    }
    outputInfo(`Deployment with id ${deploymentId} cancelled.`, logger);
    return response.deploymentCancel;
  } catch (error) {
    errorHandler(error);
    throw error;
  }
}

export { deploymentCancel };
