// @shopify/oxygen-cli/dist/deploy/deployment-complete.js
import { graphqlRequest } from '@shopify/cli-kit/node/api/graphql';
import { AbortError } from '@shopify/cli-kit/node/error';
import { authBypassTokenDurationParse, Header, errorHandler } from '../utils/utils.js';
import { DeploymentCompleteQuery } from './graphql/deployment-complete.js';

async function deploymentComplete(config, deploymentId) {
  const variables = {
    deploymentId,
    generateAuthBypassToken: config.generateAuthBypassToken
  };
  if (config.generateAuthBypassToken) {
    variables.authBypassTokenDuration = authBypassTokenDurationParse(
      config.authBypassTokenDuration
    );
  } else if (config.authBypassTokenDuration) {
    throw new AbortError(
      "authBypassTokenDuration is set but generateAuthBypassToken is not enabled"
    );
  }
  if (config.overriddenEnvironmentVariables) {
    variables.environmentVariables = config.overriddenEnvironmentVariables;
  }
  try {
    const response = await graphqlRequest({
      query: DeploymentCompleteQuery,
      api: "Oxygen",
      url: `${config.deploymentUrl}/api/v2/admin/graphql`,
      token: config.deploymentToken.accessToken,
      variables,
      addedHeaders: {
        [Header.OxygenNamespaceHandle]: config.deploymentToken.namespace
      }
    });
    if (response.deploymentComplete.userErrors.length >= 1) {
      throw new AbortError(
        `Failed to complete deployment: ${response.deploymentComplete.userErrors[0]?.message}`
      );
    }
    return response.deploymentComplete;
  } catch (error) {
    errorHandler(error);
    throw error;
  }
}

export { deploymentComplete };
