// @shopify/oxygen-cli/dist/deploy/verify-deployment-completed.js
import { graphqlRequest } from '@shopify/cli-kit/node/api/graphql';
import { outputInfo, outputCompleted } from '@shopify/cli-kit/node/output';
import { Header } from '../utils/utils.js';
import { DeploymentCompletedVerificationError } from './types.js';
import { DeploymentVerificationDetailsQuery, Status } from './graphql/deployment-verification-details.js';

async function verifyDeploymentCompleted(options, deploymentId) {
  const { config, logger, timeoutInSeconds, hooks } = options;
  const variables = {
    deploymentId
  };
  hooks?.onDeploymentCompletedVerificationStart?.();
  outputInfo("Verifying deployment completed...", logger);
  const startTime = Date.now();
  const handleInterval = async () => {
    const delay = 3e3;
    const elapsedTime = (Date.now() - startTime) / 1e3;
    if (elapsedTime + delay / 1e3 > timeoutInSeconds) {
      hooks?.onDeploymentCompletedVerificationError?.();
      throw new Error(
        `Deployment not completed after ${timeoutInSeconds} seconds.`
      );
    }
    await new Promise((resolve) => setTimeout(resolve, delay));
    return check();
  };
  const check = async () => {
    const response = await graphqlRequest({
      query: DeploymentVerificationDetailsQuery,
      api: "Oxygen",
      url: `${config.deploymentUrl}/api/v2/admin/graphql`,
      token: config.deploymentToken.accessToken,
      variables,
      addedHeaders: {
        [Header.OxygenNamespaceHandle]: config.deploymentToken.namespace
      }
    });
    const data = response.deploymentVerificationDetails;
    if (data == null) {
      hooks?.onDeploymentCompletedVerificationError?.();
      throw new DeploymentCompletedVerificationError("Deployment not found");
    }
    if (data.status === Status.Deployed) {
      outputCompleted("Deployment to Oxygen completed successfully", logger);
      hooks?.onDeploymentCompleted?.();
      return data;
    }
    if (data.status === Status.Failed || data.status === Status.Cancelled) {
      hooks?.onDeploymentFailed?.(data);
      throw new DeploymentCompletedVerificationError(
        data.error ? `Deployment ${data.status.toLocaleLowerCase()}, error: ${data.error}` : `Deployment ${data.status.toLocaleLowerCase()}`
      );
    }
    return handleInterval();
  };
  return check();
}

export { verifyDeploymentCompleted };
