// @shopify/oxygen-cli/dist/utils/bugsnag.js
import Bugsnag from '@bugsnag/node';
import { loadPackageJson } from './utils.js';

let BugsnagClient;
function setupBugsnag(disableBugsnag) {
  if (disableBugsnag) {
    return;
  } else if (Bugsnag._client) {
    return Bugsnag;
  }
  const packageJson = loadPackageJson();
  const bugsnagServerOptions = {
    apiKey: "21fd37fe87ead4b9c5a8eed90752b455",
    autoDetectErrors: true,
    appVersion: packageJson.version,
    logger: null
  };
  return Bugsnag.start(bugsnagServerOptions);
}
function initializeBugsnag(disableBugsnag) {
  BugsnagClient = setupBugsnag(disableBugsnag);
}
function getBugsnag() {
  return BugsnagClient;
}
function sanitizeError(error, rootPath) {
  if (!rootPath) {
    return error;
  }
  const replaceRootPath = (message) => {
    return message.replace(new RegExp(rootPath, "g"), "***");
  };
  if (error instanceof Error) {
    const cleanError = Object.create(Object.getPrototypeOf(error));
    Object.assign(cleanError, error, {
      message: replaceRootPath(error.message),
      stack: error.stack ? replaceRootPath(error.stack) : void 0
    });
    return cleanError;
  }
  return replaceRootPath(error);
}

export { getBugsnag, initializeBugsnag, sanitizeError };
